engine/notifications.js

```diff
@@ -0,0 +1,298 @@
+import redis from '../engine/redis-helper.js';
+import webPush from 'web-push';
+
+const ttl = 2419200;
+
+if (!process.env.GCM_API_KEY) {
+  console.warn('Set the GCM_API_KEY environment variable to support GCM');
+}
+webPush.setGCMAPIKey(process.env.GCM_API_KEY);
+
+// there is only one client (useful especially while testing)
+let client;
+function setClient(dbNumber) {
+  if (client) {
+    return Promise.resolve(client);
+  }
+  return redis.getClient(dbNumber)
+  .then(redisClient => {
+    client = redisClient;
+    return client;
+  });
+}
+
+// eported only for test as we will not need to kill database
+function quitClient(dbNumber) {
+  return setClient(dbNumber)
+  .then(() => {
+    const p = redis.quit(client);
+    client = null;
+    return p;
+  });
+}
+
+// check if such device exists in database, reject if not
+function checkDeviceId(deviceId, dbNumber) {
+  return setClient(dbNumber)
+  .then(() => redis.hgetall(client, `device-${deviceId}`))
+  .then(device => {
+    if (!device || !device.endpoint) {
+      throw new Error('Not Found');
+    }
+  });
+}
+
+// get all registrations for a deviceId
+function getRegisteredFeatures(deviceId, dbNumber) {
+  return checkDeviceId(deviceId, dbNumber)
+  .then(() => redis.smembers(client, `${deviceId}-notifications`));
+}
+
+/*
+ * send confirmation even if deviceId is already deleted
+ */
+function sendConfirmation(endpoint, key, deviceId, features, dbNumber) {
```

Comment:
Instead of passing `endpoint`, `key` and, soon, `authSecret`, you could create and pass around a `device` object with those three as properties.